### ISSUE TITLE
Redesigning the executor

### ISSUE THREAD

**Issue by stjepang** [stjepang/smol#202]
Some new ideas for executor design:
- I am considering removing `Task::spawn()`. The whole "inside vs outside the executor" concept is really confusing and it needs to go. Here's a bit heated opinion of someone trying out async for the first time: https://twitter.com/khuey_/status/1291187035588313088
- smol could expose `Executor` and `LocalExecutor` types from `async-executor` so you'd be encouraged to create your own executor and then explicitly choose the executor to spawn futures onto: `executor.spawn(future)`.
- `Executor` could have a const constructor so you could easily define a global one with `static EX: Executor = Executor::new()`
- `Executor` can implement the `Spawner` trait.
- Still, sometimes it's nice to have a default global executor so that you don't have to create one yourself. This is good for simple programs but not for libraries, i think. So I'd make a default global `Executor` and have `smol::spawn()` to spawn futures onto it.

Things I'm not sure about:
- Should there be `smol::run()` and if so, what should it do exactly? Is it just a simple `block_on()` then? Should it create a simple thread-local executor onto which you could spawn with `smol::spawn_local()`?
- How many threads should the default global executor create? I'm torn between `num_cpus::get()` and 1.
- What happens when a future on the global executor panics? Propagate the panic into `task.await`? Or just abort the process?

To summarize, the goals of this redesign would be:
- Remove the concept of "inside" and "outside" the executor. `Task::spawn()` and `Task::local()` will go away.
- Make it very easy to create instances of executors on your own.
- Make it possible to easily spawn a future *somewhere* in simple programs and unit tests. You often don't care where the future goes, it just needs to run on some kind of global executor.
- It would also be nice to make it possible to create a simple, isolated executor with a handy function like `smol::run()`. But that is optional, I think...

**Comment by serzhiio** [comment]
> Should there be smol::run() and if so, what should it do exactly? Is it just a simple block_on() then? Should it create a simple thread-local executor onto which you could spawn with smol::spawn_local()?
::block_on() on current thread

> How many threads should the default global executor create? I'm torn between num_cpus::get() and 1.
current thread, all other options should be user-defined with choosen executor

> What happens when a future on the global executor panics? Propagate the panic into task.await? Or just abort the process?
i like global panic variant but propagating to task.await is also acceptable

**Comment by oblique** [comment]
> Some new ideas for executor design:
>
> * I am considering removing `Task::spawn()`. The whole "inside vs outside the executor" concept is really confusing and it needs to go. Here's a bit heated opinion of someone trying out async for the first time: https://twitter.com/khuey_/status/1291187035588313088
> * smol could expose `Executor` and `LocalExecutor` types from `async-executor` so you'd be encouraged to create your own executor and then explicitly choose the executor to spawn futures onto: `executor.spawn(future)`.
> * `Executor` could have a const constructor so you could easily define a global one with `static EX: Executor = Executor::new()`
> * `Executor` can implement the `Spawner` trait.
> * Still, sometimes it's nice to have a default global executor so that you don't have to create one yourself. This is good for simple programs but not for libraries, i think. So I'd make a default global `Executor` and have `smol::spawn()` to spawn futures onto it.
>

This is nice. Some questions:

What will happen when someone calls `smol::spawn` when is using `async-std`? Two `Executor` instances will be initialized?

As a library developer I think it will use `async-executor` instead. Personally I don't mind that if you give me a way to check if `Executor` is already initialized or not, so I will start mine. For example async-executor can have a `Task::try_spawn`. What I'm looking is a way to write an independent library that will spawn an executor thread only if needed.

Another question is: Will async-executor run both? `Executor` and `LocalExecutor`? (https://github.com/stjepang/async-executor/pull/4)

> * How many threads should the default global executor create? I'm torn between `num_cpus::get()` and 1.

Can this be configurable and if someone doesn't configure it, then use `num_cpus::get()`?

> * What happens when a future on the global executor panics? Propagate the panic into `task.await`? Or just abort the process?

Propagation is nicer, since someone can catch it with `catch_unwind`.

**Comment by stjepang** [comment]
> What will happen when someone calls `smol::spawn` when is using `async-std`? Two `Executor` instances will be initialized?

Yes, two instances would be initialized. This is difficult to avoid, unfortunately, as async-std has a bit different executor design due to the use of task-locals.
> Another question is: Will async-executor run both? `Executor` and `LocalExecutor`? ([stjepang/async-executor#4](https://github.com/stjepang/async-executor/pull/4))

I'm not sure. My current feeling is that bundling `Executor` and `LocalExecutor` together might be more confusing than it is useful. But maybe it isn't, I don't know.

> Can this be configurable and if someone doesn't configure it, then use `num_cpus::get()`?

It could be. I think the choice is between:

* If the default is `num_cpus::get()`, then people will treat it as a "real" executor similar to the one in async-std. The danger here is that smol feels bloated and may spawn more threads than is actually needed.
* If the default is 1, then this global executor is there just to "help" as a fallback thread for "lost" futures that don't know where to go.

> As a library developer I think it will use `async-executor` instead. Personally I don't mind that if you give me a way to check if `Executor` is already initialized or not, so I will start mine. For example async-executor can have a `Task::try_spawn`. What I'm looking is a way to write an independent library that will spawn an executor thread only if needed.

This may be a bigger problem that is out of scope of smol. We need a widely accepted unified abstraction (like the `Spawn` trait) that is embraced by all major executors. We can check whether `Executor` is already initialized, but I think in reality it's more likely the current executor is `async-std` or `tokio` rather than `Executor`.

**Comment by oblique** [comment]
> > Another question is: Will async-executor run both? `Executor` and `LocalExecutor`? ([stjepang/async-executor#4](https://github.com/stjepang/async-executor/pull/4))
> 
> I'm not sure. My current feeling is that bundling `Executor` and `LocalExecutor` together might be more confusing than it is useful. But maybe it isn't, I don't know.

In reality I think it will be very difficult to spawn a non-Send future without this. Imagine  being in a nested `Task` and you want to spawn a non-Send future.

> It could be. I think the choice is between:
> 
> * If the default is `num_cpus::get()`, then people will treat it as a "real" executor similar to the one in async-std. The danger here is that smol feels bloated and may spawn more threads than is actually needed.
> * If the default is 1, then this global executor is there just to "help" as a fallback thread for "lost" futures that don't know anywhere else to go.

In that case, the 1 as default is preferable.

**Comment by sehz** [comment]
Seems like executor should be treated like [allocator](https://doc.rust-lang.org/alloc/alloc/trait.GlobalAlloc.html).   It should be configured at startup depends on application use case.

**Comment by redtankd** [comment]
async-executor, multitask, async-task all have Task, which seems a little confused. Task in async-executor and multitask is a wrap of JoinHandle actually. Why not re-export JoinHandle simply ?

**Comment by stjepang** [comment]
@redtankd I don't like the `JoinHandle` interface (that's what `async-task` has), and prefer `Task` that cancels the task when dropped. I'm also considering rewriting the `async-task` crate with less unsafe code...

I agree that `async-executor` doesn't need its own `Task` type. I admit this is all quite confusing :)

**Comment by ian-p-cooke** [comment]
so recently I wanted to test a new library of mine with the popular executors to see what the performance was like.  I ran into this inside/outside executor problem because I had a library which should use whatever executor the application wanted.  I wanted to switch executors with one line of code or config.  While thinking about this problem and after talking about it a little I came up with this idea and ran with it for a little bit.

https://github.com/ian-p-cooke/malebolgia

The idea is that there's some top-level function to spawn futures into that's set by an application.  As application author then needs to pick a runtime, depends on malebolgia and select a compatibility feature for the runtime.  You end up with one more dependency that only builds support for the runtime you wanted and have to add one line of code to your application.  seems like a win to me but what do you think?

Most of the executors, with the exception of Futures 0.3 have a facility for spawning into some global runtime.  However, I kind of like that they didn't do that because then it's super-clear that you need to set up your runtime, set it for spawning, and then go.  smol could do the same thing.  get rid of the spawning functions and just have spawning methods on the runtime.  Leave the 'use some global executor for library spawning' up to some other crate like malebolgia.

I didn't put my code up on crates.io because it's still just an idea to me and needs to be fleshed out through actual usage.  But the spawning does work as expected now so you can play with that using a git dependency.  Beyond spawning I added the things I thought every spawn-facility should provide which is spawning a (blocking) function and canceling/detaching the spawned future/function.

**Comment by joshtriplett** [comment]
I like the current behavior of `smol::run`, starting a thread per CPU. If that stops being the behavior of `smol::run`, then I'd like to have a function equivalent to the current `smol::run`, rather than having to reinvent it. I'd be fine with an explicit `smol::run_threaded`, to indicate that yes, I really want threads.

**Comment by arzg** [comment]
I haven’t used smol before, so feel free to completely disregard this comment, but I just wanted to chime in that [structured concurrency](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/) might be worth considering. I guess you’d have an API roughly like this:

```rust
impl Executor {
    async fn scope<R>(&mut self, f: impl Fn(Scope) -> impl Future<Output=R>) -> R;
}

struct Scope {
    ...
}

impl Scope {
    async fn spawn<R>(&mut self, task: impl Fn() -> impl Future<Output=R>) -> R;
}
```

that can be used like this:

```rust
let mut executor = Executor::new();

executor.scope(|scope| {
    scope.spawn(my_async_fn);
    scope.spawn(my_other_async_fn);

    // The executor joins on both of these tasks here, so no tasks can ‘escape’ the scope.
});
```

**Comment by stjepang** [comment]
@argz Scoped executors should be possible soon. I'm not quite there but *almost*...... :)

**Comment by stjepang** [comment]
I have published smol v0.4.0. Notable changes:

* `smol::block_on()` now simply blocks on a future.
* `smol::run()` is removed.
* crates async-lock, async-executor, async-channel, async-net, async-fs, async-process are now re-exported (this adds very little, almost nothing to the compile times).
* Type `Executor` can be used to create executors, its `run()` method is now an async fn, and `new()` is a const fn.
* `unblock!()` macro is removed in favor of the `unblock()` function.
* `smol::spawn()` spawns onto a global single-threaded executor, but it's easy to create your own executors.
* `Task::spawn()` and `Task::local()` are removed - use `Executor::spawn()` and `LocalExecutor::spawn()` instead.

**Comment by stjepang** [comment]
As a fun proof of concept, here's an executor with three task priorities (high, medium, low):

https://github.com/stjepang/async-executor/blob/master/examples/priority.rs

It's surprisingly simple to implement!

**Comment by redtankd** [comment]
Maybe some default executors be good for newcomers.

**Comment by benmkw** [comment]
> As a fun proof of concept, here's an executor with three task priorities (high, medium, low):
> 
> https://github.com/stjepang/async-executor/blob/master/examples/priority.rs
> 
> It's surprisingly simple to implement!

Really cool to see this, I thought this might be nice on embedded as well, especially on avr where RTIC is not currently available. (I know it’s not a replacement but it would offer some options for scheduling) RTIC wants to have an executor for the idle task as well iirc so it could even be interesting there for finer granularity.

**Comment by stjepang** [comment]
@redtankd 

> Maybe some default executors be good for newcomers.

Could you say more? What would be an ideal API to you? Would it help to include the following examples in the docs?

```rust
static EX: smol::Executor = smol::Executor::new();

smol::block_on(EX.run(async {
    let task = EX.spawn(async { 1 + 2 });
    assert_eq!(task.await, 3);
});
```

```rust
static EX: smol::Executor = smol::Executor::new();

for _ in 0..num_cpus::get() {
    let run = EX.run(smol::future::pending::<()>());
    std::thread::spawn(move || smol::block_on(run));
}
```

Or, perhaps as another idea, maybe we could add the following new API as a thin layer over `Executor`:

```rust
struct ThreadPool;

impl ThreadPool {
    const fn fn new() -> ThreadPool;

    const fn fn with_size(num_threads: usize) -> ThreadPool;

    fn spawn<T: Send + 'static>(
        f: impl Future<Output = T> + Send + 'static,
    ) -> Task<T>;
}
```

**Comment by arzg** [comment]
I personally would love to see an executor that can _only_ spawn tasks through scopes, so that leaked tasks are impossible to create. Is there anyone else here that would like this?

**Comment by stjepang** [comment]
@arzg Do you mean something like this? Because that is totally possible to implement:

```rust
struct ScopedExecutor<'a>;

impl<'a> ScopedExecutor<'a> {
    fn new() -> ScopedExecutor<'a>;

    fn spawn<T: Send + 'a>(
        &self,
        f: impl Future<Output = T> + Send + 'a,
    ) -> Task<T>;

    async fn run<T>(&self, impl Future<Output = T>) -> T;
}
```

**Comment by arzg** [comment]
What I meant is something like the API I outlined in [my earlier comment](https://github.com/stjepang/smol/issues/202#issuecomment-679963323):

> ```rust
> impl Executor {
>     async fn scope<R>(&mut self, f: impl Fn(Scope) -> impl Future<Output=R>) -> R;
> }
>
> struct Scope {
>     ...
> }
>
> impl Scope {
>     async fn spawn<R>(&mut self, task: impl Fn() -> impl Future<Output=R>) -> R;
> }
> ```

The specific thing I was mentioning is that `Executor` would _not_ have any way of spawning tasks, apart from `scope`. This would prevent the creation of leaked tasks.

Removing any other method of creating tasks is sort of like removing `goto`: less powerful, but prevents errors and makes control flow clearer.

**Comment by stjepang** [comment]
I'm not sure I understand - what exactly are leaked tasks in this context?

Is the idea that at the end of `scope()`, all tasks spawned inside it are canceled, i.e. dropped? If so, isn't that equivalent to just creating an `Executor` at the beginning of scope and then dropping it at the end?

**Comment by arzg** [comment]
> I'm not sure I understand - what exactly are leaked tasks in this context?

Leaked tasks are tasks that aren’t joined, and just keep running as long as they can (until the executor is dropped).

> Is the idea that at the end of `scope()`, all tasks spawned inside it are canceled, i.e. dropped?

No, the idea is that all tasks spawned on a `Scope` are joined when the `Scope` is dropped.

**Comment by stjepang** [comment]
Ah, I understand now, thanks! Will need to think a bit more about this. It should be easy to implement your proposed API using `Executor` right now, actually.

**Comment by arzg** [comment]
That’s good to hear! Thanks for all your work on the async ecosystem :)

**Comment by andrewbanchich** [comment]
Maybe `async_executor` could have a collection of different Executor implementations like `PriorityExecutor` people can either use or reference to build their own.

**Comment by stjepang** [comment]
@andrewbanchich I'm a bit hesitant to add more executors beside `Executor` and `LocalExecutor` because the possibilities are seemingly endless and those two executors are kind of the basic building block for making new executors.

I think I'd put more interesting executors in a new crate...

**Comment by andrewbanchich** [comment]
Sounds great! BTW, I migrated to 0.4 today and am really enjoying this new API. Having `Executor` be static really simplified a lot.

Is `CompatExt` something libraries can use to wrap "leaf incompatibilities" like `reqwest` calls and then they can market themselves as runtime agnostic, or is it more for users that already have an executor running?

**Comment by stjepang** [comment]
> Is `CompatExt` something libraries can use to wrap "leaf incompatibilities" like `reqwest` calls and then they can market themselves as runtime agnostic, or is it more for users that already have an executor running?

Eh, depends on who you ask :) Marketing can be pretty controversial - I have been burnt by this before am now trying to stay out of it. The main confusion around the word "runtime-agnostic" is on whether we mean runtime-agnostic on the inside or on the outside?

I tend to look at runtime-agnostic crates from the outside. Is the experience of using them exactly the same no matter what runtime you're using? If yes, then cool, it is runtime-agnostic.

But some people look at it from the inside. Is it running a runtime in the background, or pulling in a specific runtime as a dependency? If yes, then they say it's not runtime-agnostic. In a way, it's "cheating" by pretending it doesn't require a runtime because it's running one on its own.

Another way of looking at it is: does the crate force you to pick a specific runtime, or make your life difficult unless you choose a specific runtime? That's what it means to look at it from the outside and is the definition of "runtime-agnostic" I am personally using.

**Comment by stjepang** [comment]
The executor has been successfully redesigned so closing this issue. Thanks everyone for the feedback, ideas, and the discussion!
